JSIL.Libraries/Sources/JSIL.Core.js

```diff
@@ -8367,50 +8522,170 @@ JSIL.ResolvedMethodSignature.prototype.toString = function () {
     this.signature = signature.Clone(true);
   } else {
-    // FIXME: Why the hell does this happen?
     this.signature = null;
   }
 
-  this.parameterInfo = parameterInfo;
   this.qualifiedName = JSIL.$GetSignaturePrefixForType(typeObject) + this.methodName;
   this.variantInvocationCandidateCache = JSIL.CreateDictionaryObject(null);
   if (interfaceMemberFallbackMethod !== null) {
     this.fallbackMethod = interfaceMemberFallbackMethod;
   }
 
   JSIL.SetLazyValueProperty(this, "methodKey", function () {
-    return this.signature.GetNamedKey(this.qualifiedName, true);
+    return this.signature != null ? JSIL.$GetSignaturePrefixForType(typeObject) + this.methodNonQualifiedKey : this.qualifiedName;
+  });
+
+  JSIL.SetLazyValueProperty(this, "methodNonQualifiedKey", function () {
+    return this.signature != null ? this.signature.GetNamedKey(this.methodName, true) : this.methodName;
   });
 };
 
-JSIL.SetLazyValueProperty(JSIL.InterfaceMethod.prototype, "Call", function () { return this.$MakeCallMethod(); }, true);
+JSIL.SetLazyValueProperty(JSIL.InterfaceMethod.prototype, "Call", function () { return this.$MakeCallMethod(true); }, true);
+JSIL.SetLazyValueProperty(JSIL.InterfaceMethod.prototype, "CallNonVirtual", function () { return this.$MakeCallMethod(false); }, true);
 
 JSIL.InterfaceMethod.prototype.Rebind = function (newTypeObject, newSignature) {
-  var result = new JSIL.InterfaceMethod(newTypeObject, this.methodName, this.signature != null ? newSignature : null, this.parameterInfo);
-  result.fallbackMethod = this.fallbackMethod;
+  if (this.__OfCache__ == null) {
+    JSIL.RuntimeError("Rebind of signature-aware InterfaceMethod");
+  }
+
+  var result = new JSIL.InterfaceMethod(newTypeObject, this.methodName, newSignature, this.fallbackMethod);
+  return result;
+};
+
+JSIL.InterfaceMethod.prototype.RegisterSignature = function (signature, skipExistingCheck) {
+  if (this.__OfCache__ == null) {
+    JSIL.RuntimeError("RegisterSignature of signature-aware InterfaceMethod");
+  }
+  if (signature == null) {
+    JSIL.RuntimeError("InterfaceMethod RegisterSignature should accept non-null signature");
+  }
+
+  if (this.signature != null) {
+    var selfSignature = this.signature;
+    this.signature = null;
+    this.RegisterSignature(selfSignature);
+  }
+
+  var cacheKey = signature.get_Hash();
+
+  var result = this.__OfCache__[cacheKey];
+  if (result) {
+    if (skipExistingCheck) {
+      return;
+    }
+
+    JSIL.RuntimeErrorFormat(
+    "InterfaceMethod '{0}' already has registred signature {1}", [
+      this.methodName,
+      signature.toString(this.methodName)]);
+  }
+
+  var signatureAwareInterfaceMethod = new JSIL.InterfaceMethod(this.typeObject, this.methodName, signature, this.fallbackMethod);
+  signatureAwareInterfaceMethod.__OfCache__ = null;
+
+  this.__OfCache__[cacheKey] = signatureAwareInterfaceMethod;
+};
+
+JSIL.InterfaceMethod.prototype.Of = function (signature) {
```

Comment:
We can get InterfaceMethod with specific signature, that was previously registered inside. It will be used instead of calling Signature with passing InterfaceMethod as a name.